File: utils/help.py
import os
import pandas as pd
import seaborn as sns


def read_csv(path: str) -> dict:
    """
    Loads all .csv files present in path into a dictionary
    """
    csv_dict = {}
    for file in os.listdir(path):
        prefix, suffix = file.split(".")
        if suffix == "csv":
            csv_dict[prefix] = pd.read_csv(os.path.join(path, file))
    return csv_dict


def read_csv_v2(path: str) -> dict:
    """
    Just like read_csv but removes padding
    """
    csv_dict = {}
    for file in os.listdir(path):
        prefix, suffix = file.split(".")
        if suffix == "csv":
            csv_dict[str(int(prefix))] = pd.read_csv(os.path.join(path, file))
    return csv_dict


def read_csv_v3(path: str) -> dict:
    """
    Removes padding using another function
    """
    csv_dict = {}
    for file in os.listdir(path):
        prefix, suffix = file.split(".")
        if suffix == "csv":
            csv_dict[remove_padding(prefix)] = pd.read_csv(os.path.join(path, file))
    return csv_dict


def read_csv_v4(path: str, id_list=None) -> dict:
    """
    Makes it possible to load only a subset of the data
    """
    csv_dict = {}
    files_listed = [f"{station.zfill(3)}.csv" for station in id_list]
    for file in os.listdir(path):
        prefix, suffix = file.split(".")
        if suffix == "csv":
            if id_list is None or file in files_listed:
                csv_dict[remove_padding(prefix)] = pd.read_csv(os.path.join(path, file))
    return csv_dict


def test_read_csv(path):
    """
    Test for the read_csv method
    """
    dict_full = read_csv(path)
    assert len(dict_full) == len([f for f in os.listdir(path) if ".csv" in f])
    print(f"Tests passed for the test_read_csv({path})")


def remove_padding(tgt_string: str) -> str:
    """
    Removes the padding from the file
    """
    str_out = str(int(tgt_string))
    return str_out


def test_remove_padding(test_cases=['0125', '1250']):
    assert remove_padding(test_cases[0] == '125')
    print(f"remove_pading({test_cases[0]}) -> {remove_padding(test_cases[0])} Test OK")
    assert remove_padding(test_cases[1] == '1250')
    print(f"remove_pading({test_cases[1]}) -> {remove_padding(test_cases[1])} Test OK")


def plot_monitoring_station(df, id_vars=['Date', 'ID'], value_vars=['sulfate', 'nitrate']):
    data_plotted = df.melt(id_vars=id_vars, value_vars=value_vars)
    g = sns.FacetGrid(data_plotted, col='ID', hue='variable', legend_out=True)
    g.map(sns.lineplot, 'Date', 'value')


def get_monitor(station_ids: list, dropna=True, plot=True):
    stations_data = read_csv_v4("./specdata", station_ids)
    stations_data = pd.concat(stations_data.values())
    if dropna:
        stations_data = stations_data.dropna()
    if plot:
        plot_monitoring_station(stations_data)
    return stations_data

File: utils/help.R
library(readr)
library(dplyr)
library(tidyr)
library(ggplot2)

read_csv <- function(path) {
  # Loads all .csv files present in path into a list
  csv_list <- list()
  files <- list.files(path)
  for (file in files) {
    if (grepl("\\.csv$", file)) {
      prefix <- sub("\\.csv$", "", file)
      csv_list[[prefix]] <- read_csv(file.path(path, file))
    }
  }
  return(csv_list)
}

read_csv_v2 <- function(path) {
  # Just like read_csv but removes padding
  csv_list <- list()
  files <- list.files(path)
  for (file in files) {
    if (grepl("\\.csv$", file)) {
      prefix <- sub("\\.csv$", "", file)
      csv_list[[as.numeric(prefix)]] <- read_csv(file.path(path, file))
    }
  }
  return(csv_list)
}

remove_padding <- function(tgt_string) {
  # Removes the padding from the file
  str_out <- as.numeric(tgt_string)
  return(str_out)
}

read_csv_v3 <- function(path) {
  # Removes padding using another function
  csv_list <- list()
  files <- list.files(path)
  for (file in files) {
    if (grepl("\\.csv$", file)) {
      prefix <- sub("\\.csv$", "", file)
      csv_list[[remove_padding(prefix)]] <- read_csv(file.path(path, file))
    }
  }
  return(csv_list)
}

read_csv_v4 <- function(path, id_list=NULL) {
  # Makes it possible to load only a subset of the data
  csv_list <- list()
  files_listed <- paste(sprintf("%03d", id_list), ".csv", sep="")
  files <- list.files(path)
  for (file in files) {
    if (grepl("\\.csv$", file)) {
      if (is.null(id_list) | file %in% files_listed) {
        prefix <- sub("\\.csv$", "", file)
        csv_list[[remove_padding(prefix)]] <- read_csv(file.path(path, file))
      }
    }
  }
  return(csv_list)
}

plot_monitoring_station <- function(df, id_vars=c("Date", "ID"), value_vars=c("sulfate", "nitrate")) {
  # Plots the monitoring station data
  data_plotted <- pivot_longer(df, cols=value_vars, names_to="variable", values_to="value")
  g <- ggplot(data_plotted, aes(x=Date, y=value, color=variable)) +
       geom_line() +
       facet_wrap(~ ID, ncol=2) +
       theme(legend.position="bottom")
  print(g)
}

get_monitor <- function(station_ids, dropna=TRUE, plot=TRUE) {
  # Returns the monitoring station data for a list of station IDs
  stations_data <- read_csv_v4("./specdata", station_ids)
  stations_data <- bind_rows(stations_data)
  if (dropna) {
    stations_data <- stations_data %>% drop_na()
  }
  if (plot) {
    plot_monitoring_station(stations_data)
  }
  return(stations_data)
}